### PR TITLE
Cfy 2256 document security

### PR DESCRIPTION
**Pull request by noakup**

### CODE REVIEW

File: source/guide/3.2/guide-security.md [hunk 1]
```diff
@@ -11,33 +11,230 @@ pageord: 500
 
 # Overview
 
-This guide will quickly explain how to bootstrap a secured manager and use if from the cli and web UI.
+This guide will explain how to bootstrap a secured manager and use it from the CLI and web UI. <br/>
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: fyi, <br> is enough, the slash is not needed

File: source/guide/3.2/guide-security.md [hunk 2]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+Securing the manager focuses on the REST service, which is the access point of all clients to the management server.
+
+Cloudify's security framework uses the [Flask-secuREST](https://github.com/cloudify-cosmo/flask-securest/tree/0.6)
+package, which integrates with [Flask-RESTful](https://flask-restful.readthedocs.org/en/0.3.2/) to secure REST services.
+
+When security is enabled, all requests to the REST service are authenticated before they can reach their endpoint.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: some requests aren't authenticated if im not mistaken (version?)

File: source/guide/3.2/guide-security.md [hunk 3]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+For example, when a user of the Web-UI attempts to see the all blueprints, a request is sent to the REST service
+"blueprints" endpoint, but it will only reach it if the user is logged in with valid credentials. <br/>
+Similarly, the CLI command "cfy deployments create" will only be successful if valid
+username and password are sent with the REST call to the endpoint "deployments".
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: "username and password" --> "authentication information" or whatever, since it can be a token as well

File: source/guide/3.2/guide-security.md [hunk 4]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+To support a variety of user-store systems and configurations Cloudify security framework can accept different
+userstore implementations. It's possible to use the default Flask-secuREST simple userstore or to specify a new
+implementation that supports a specific userstore system.
+
+A valid userstore implementation can be any Python class that inherits from
+[abstract_userstore.py](https://github.com/cloudify-cosmo/flask-securest/blob/master/flask_securest/userstores/
+abstract_userstore.py) and implements a `get_user` method.
+The object returned by `get_user` must adhere to Flask-secuREST's User Model, specifying these 3 methods:
+
+- is_active()
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: probably requires more information on these here.. or a link to someplace else which has it

File: source/guide/3.2/guide-security.md (hunk 3 above)
Comment (noakup) [inline source/guide/3.2/guide-security.md]: not precise, since the CLI only uses user-password auth and not tokens, that's why I wrote it that way

File: source/guide/3.2/guide-security.md [hunk 5]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+- is_anonymous()
+- get_roles()
+
+
+## Authentication Provider
+An Authentication Provider is a class that performs authentication. Multiple authentication providers can be configured
+in order to support different authentication methods (e.g. password, token, Kerberos).<br/>
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: different --> multiple
makes it clearer that they're all used simultaneously

File: source/guide/3.2/guide-security.md [hunk 6]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+authentication provider in the order of configuration, until authentication succeeds.
+The authentication provider has access to the userstore instance (if configured) and can use it to get user details and
+use them to perform authentication.<br/>
+For example, it can compare the given password to the one found on the userstore or verify the user is still active
+(in many environments users are marked as "inactive", instead of deleting the account entirely).
+
+
+Once an authenticator can successfully authenticate the request's user - it should return the user object and allow the
+request to be completed. Other authenticators will not be called until the next request is processed.
+If non of the authenticators authenticates the request - the request will not reach its endpoint and the client will
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: non --> none,
authenticates --> authenticate

File: source/guide/3.2/guide-security.md [hunk 7]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+receive an "Unauthorized User" error.
+
+A valid authentication provider implementation can be any Python class that inherits from
+[abstract_authentication_provider.py](https://github.com/cloudify-cosmo/flask-securest/blob/master/flask_securest/
+authentication_providers/abstract_authentication_provider.py) and implements an `authenticate` method.
+
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: kinda unclear about the signature of the method itself - from what i understand, if i want to return a success, then the method should return the user object. what i should i do otherwise? raise an error? return false or none?

File: source/guide/3.2/guide-security.md [hunk 8]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+## Token Generator
+We mentioned Token as an authentication method. But in order to send a token with each request, the user must first
+receive a token. Tokens can be generated by many systems, and they will work as long as the token can be processed by
+one of the registered authentication providers.
+
+
+To make things easier, Cloudify can also generate tokens through the REST service endpoint "/tokens".
+To enable this feature a token generator must be configured.
+{%note title=Note%}
+The request to "MANAGER_IP/tokens" must itself be authenticated (using a username-password set, for example).
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: "MANAGER_IP/tokens" --> "/tokens" endpoint
more consistent with what you've been going for so far

File: source/guide/3.2/guide-security.md [hunk 9]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+{%endnote%}
 
-## Authentication Providers
 
 # Setting up a secured server
 
 ## Manager Blueprint Configuration
+The security configuration is located in the manager blueprint, in this path:
+node_templates --> manager --> properties --> cloudify --> security
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: should probably use dots instead of arrows, as its the standard notation on all of our docs

File: source/guide/3.2/guide-security.md [hunk 10]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+
+{% highlight yaml %}
+enabled: false
+{% endhighlight %}
+
+This means security is turned off. In order to activate the security framework set {% highlight yaml %}enabled: true{% endhighlight %}
+Otherwise, all other security configuration will be ignored.
+
+### Configuring Authentication Providers
+In the manager blueprint, under `authentication_providers` is a list all the authenticators *In the order they should be executed*.<br/>
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: the "in the manager blueprint" here actually threw me off; basically this is a rundown of the config in the manager blueprint. we started with the first config setting which was setting security on/off, and so this is the direct continuation of that right? anyway i'd drop the first part, but maybe thats just me getting confused easily

File: source/guide/3.2/guide-security.md [hunk 11]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+    implementation: flask_securest.authentication_providers.token:TokenAuthenticator
+    properties:
+      secret_key: my_secret
+{% endhighlight %}
+
+The above configuration will cause the security framework to instantiate two classes:
+
+- Flask-secuREST's "PasswordAuthenticator", with the password_hash arguement set to "plaintext".
+{%note title=Note%}
+Passwords are usually not store as plaintext. Set the passowrd_hash to match the hash scheme used
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: store --> stored
password_hash --> `password_hash` (also true for line 131)

File: source/guide/3.2/guide-security.md [hunk 12]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+in the selected userstore.
+Possible values are: 'bcrypt', 'des_crypt', 'pbkdf2_sha256', pbkdf2_sha512', 'sha256_crypt' and 'sha512_crypt'.
+{%endnote%}
+This authentication check will be performed first for each request sent to the REST service. If it fails, the next
+authenticator will be used.
+
+- Flask-secuREST's "TokenAuthenticator", with the secret_key arguement set to "my_secret".
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: secret_key --> `secret_key`

File: source/guide/3.2/guide-security.md [hunk 13]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+      secret_key: my_secret
+{% endhighlight %}
+
+The above configuration will cause the security framework to instantiate two classes:
+
+- Flask-secuREST's "PasswordAuthenticator", with the password_hash arguement set to "plaintext".
+{%note title=Note%}
+Passwords are usually not store as plaintext. Set the passowrd_hash to match the hash scheme used
+in the selected userstore.
+Possible values are: 'bcrypt', 'des_crypt', 'pbkdf2_sha256', pbkdf2_sha512', 'sha256_crypt' and 'sha512_crypt'.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: missing ' somewhere around pbkdf2_sha512 i think

File: source/guide/3.2/guide-security.md [hunk 14]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+{%endnote%}
+This authentication check will be performed first for each request sent to the REST service. If it fails, the next
+authenticator will be used.
+
+- Flask-secuREST's "TokenAuthenticator", with the secret_key arguement set to "my_secret".
+If (and only if) the password-based authentication failed, this authentication method will be executed as well.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: redundant line IMO, it just said the same thing a couple of lines ago

File: source/guide/3.2/guide-security.md [hunk 15]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+The secret key is used to decrypt the token sent with the request, if a token was sent.
+{%note title=Note%}
+In the default configuration, the key used to authenticate a token must be the same as the key used to generate it.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: think i'd add a few words here as to why it must be so

File: source/guide/3.2/guide-security.md [hunk 16]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+
+### Configuring a Userstore
+Under "userstore_driver" a single userstore is set:
+
+- implementation - the fully qualified name of the module implementing a userstore integration, followed by ":" and the
+class name.
+- properties - a dictionary of arguments required to instantiate the implementing class. The arguments will be passed as
+kwargs to the class' `__init__` method.
+
+The default configuration uses Flask-secuREST's simple userstore, with a list of users in-line:
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: "a list of users in-line" --> "an in-line list of users"

File: source/guide/3.2/guide-security.md [hunk 17]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+auth_token_generator:
+  implementation: flask_securest.authentication_providers.token:TokenAuthenticator
+  properties:
+    secret_key: my_secret
+    expires_in_seconds: 600
+{% endhighlight %}
+
+This configuration uses (as it must) the same secret key for token generation and for token authentication.
+This configuration however includes the additional argument "expires_in_seconds" which limits the lifetime of a token
+to 10 minutes. A token older than 10 minutes will therefore be "expired" and fail the request.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: think this should clear up whether the expire-in-time a /tokens endpoint feature or rather its specific to the default token generator

File: source/guide/3.2/guide-security.md [hunk 18]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+### Logging
+Security operations, such as authentication success or failure and user details, are logged to a dedicated log file on the management container.
+The default log configuration is:
+
+{% highlight yaml %}
+log_file: /var/log/cloudify/rest-security-audit.log
+log_level: INFO
+{% endhighlight %}
+
+Setting the log file location is pretty strait-forward.<br/>
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: strait --> straight

Comment (ran-z) [inline source/guide/3.2/guide-security.md]: though actually, i'd drop the line altogether. it doesn't help much :D

File: source/guide/3.2/guide-security.md [hunk 19]
```diff
@@ -11,33 +11,230 @@ pageord: 500
...
+- is_active()
+- is_anonymous()
+- get_roles()
+
+
+## Authentication Provider
+An Authentication Provider is a class that performs authentication. Multiple authentication providers can be configured
+in order to support different authentication methods (e.g. password, token, Kerberos).<br/>
+When a REST call is received by the REST service, the security framework will attempt to authenticate it using each
+authentication provider in the order of configuration, until authentication succeeds.
```
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: think i'd rephrase this line, the ending sounds kind of wrong

File: source/guide/3.2/guide-security.md (hunk 3 above)
Comment (ran-z) [inline source/guide/3.2/guide-security.md]: ah ok
i didnt see anywhere this is mentioned (CLI being able to use only user-password), so without knowing it, others could find this line a bit confusing perhaps? then again probably not.
might want to mention it anyway
up to you :D

Comment (noakup) [inline source/guide/3.2/guide-security.md]: you're right, I'll add it in the clients section
